### PR TITLE
New workflow for mirapay using iframes

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Adds support for MiraSecure® Iframe Plugin in packetfence
## Impacts

The billing workflow
## NEWS file entries

New Features
++++++++++++
- New billing gateway MiraSecure Iframe Plugin

Enhancements
++++++++++++

Bug Fixes
+++++++++
## Delete branch after merge

NO

### CODE REVIEW

File: conf/documentation.conf [hunk 1]
```diff
@@ -1111,6 +1111,33 @@ description=<<EOT
...
+type=text
+description=<<EOT
+The mirapay merchant id
+EOT
+
+[billing.mirapay_language]
+type=toggle
+options=EN|FR
+description=<<EOT
+The what language to display by default
```
Comment (julsemaan) [inline conf/documentation.conf]: The what language ? Is that a language composed of only the word what ?

File: conf/documentation.conf [hunk 2]
```diff
@@ -1111,6 +1111,33 @@ description=<<EOT
 The hash password for mirapay
 EOT
 
+[billing.mirapay_iframe_url]
+type=toggle
+options=https://staging.eigendev.com/MiraSecure/GetToken.php|https://ms1.eigendev.com/MiraSecure/GetToken.php|https://ms2.eigendev.com/MiraSecure/GetToken.php
+description=<<EOT
+The url to submit the request by default it is set to the test.
```
Comment (julsemaan) [inline conf/documentation.conf]: *URL

Comment (julsemaan) [inline conf/documentation.conf]: ... . By default ....

File: conf/documentation.conf [hunk 3]
```diff
@@ -1111,6 +1111,33 @@ description=<<EOT
...
+In production it should be set to a production url
```
Comment (julsemaan) [inline conf/documentation.conf:1153]: *URL

File: conf/documentation.conf [hunk 4]
```diff
@@ -1111,6 +1111,33 @@ description=<<EOT
...
+EOT
+
+[billing.mirapay_iframe_shared_secret]
+type=text
+description=<<EOT
+The shared secret for mirapay
```
Comment (julsemaan) [inline conf/documentation.conf]: *for Mirapay

File: conf/documentation.conf [hunk 5]
```diff
@@ -1111,6 +1111,33 @@ description=<<EOT
...
+EOT
+
+[billing.mirapay_iframe_merchant_id]
+type=text
+description=<<EOT
+The mirapay merchant id
```
Comment (julsemaan) [inline conf/documentation.conf]: *Mirapay

File: conf/pf.conf.defaults [hunk 6]
```diff
@@ -858,6 +858,26 @@ mirapay_terminal_id =
 mirapay_terminal_id_group =
 
 mirapay_hash_password =
+#
+# billing.mirapay_iframe_url
+#
+# The url to send all request
```
Comment (julsemaan) [inline conf/pf.conf.defaults:889]: all requests

File: conf/pf.conf.defaults [hunk 7]
```diff
@@ -858,6 +858,26 @@ mirapay_terminal_id =
...
+mirapay_iframe_shared_secret =
+#
+# billing.mirapay_iframe_merchant_id
+#
+# The mirapay merchant id
+mirapay_iframe_merchant_id =
+#
+# billing.mirapay_language
+#
+# The what language to display by default
```
Comment (julsemaan) [inline conf/pf.conf.defaults:904]: The what language is getting popular

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm [hunk 8]
```diff
@@ -256,40 +190,382 @@ sub showBilling : Private {
...
+        tier      => $session->{'tier'},
+        firstname => $session->{'firstname'},
+        lastname  => $session->{'lastname'},
+        email     => $session->{'email'},
+    };
+    $c->forward('send_billing_email' => [$transaction_infos_ref, $confirmationInfo]);
 
-Inverse inc. <info@inverse.ca>
+    # Generate the release page
+    # XXX Should be part of the portal profile
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm:234]: What does XXX stands for ? 
The PacketFence code is getting sexy ?

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm [hunk 9]
```diff
@@ -256,40 +190,382 @@ sub showBilling : Private {
...
+        lastname    => $session->{'lastname'},
+        email       => lc($session->{'email'}),
+        item        => $tier,
+        price       => $tiers_infos{$tier}{'price'},
+        description => $tiers_infos{$tier}{'description'},
+    };
+    $c->forward('processSuccessPayment', [$tier, $transaction_infos_ref]);
+}
+
+our %MirapayResponseCode = (
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm:442]: In the middle of the file ?
Also that should be read only and in a constants file

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm [hunk 10]
```diff
@@ -256,40 +190,382 @@ sub showBilling : Private {
...
+        $c->stash->{'txt_validation_error'} = $MirapayResponseCode{$response_code};
+        return $FALSE;
+    }
+    return $TRUE;
+}
+
+
+=head2 mirapay_iframe_verify_input
+
+Verify
```
Comment (julsemaan) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Pay.pm]: I think no doc is better than only 1 word

File: html/pfappserver/lib/pfappserver/I18N/i_default.po [hunk 11]
```diff
@@ -4165,6 +4181,12 @@ msgstr ""
 msgid "The terminal id group for mirapay"
 msgstr ""
 
+# conf/documentation.conf (billing.mirapay_iframe_url)
+msgid ""
+  "The url to submit the request by default it is set to the test."
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/I18N/i_default.po]: Not sure these are valid. I may be wrong though

**Comment by jrouzierinverse** [comment]
This has been deprecated by #826

**Pull request closed without merging**